### PR TITLE
Implement conversion between `java.lang.{Byte, Short, Integer, Long}` and JString

### PR DESCRIPTION
**Pull request by t-hiroyoshi**
target: #160 
ref: #173

### CODE REVIEW

File: javalib/src/main/scala/java/lang/Long.scala [hunk 1]
```diff
@@ -36,81 +48,317 @@ object Long {
...
 
-  @inline def toString(i: scala.Long): String = ???
+  @inline def compare(x: scala.Long, y: scala.Long): scala.Int =
+    if (x == y) 0
+    else if (x < y) -1
+    else 1
 
-  @inline def toUnsignedString(i: scala.Long): String = ???
+  @inline def compareUnsigned(x: scala.Long, y: scala.Long): scala.Int =
+    undefined
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Long.scala]: Not an intrinsic, needs to be implemented in the library code.

File: javalib/src/main/scala/java/lang/Long.scala [hunk 2]
```diff
@@ -36,81 +48,317 @@ object Long {
...
+  @inline def parseLong(s: String): scala.Long =
+    parseLong(s, 10)
+
+  @inline def parseUnsignedLong(s: String): scala.Long =
+    parseUnsignedLong(s, 10)
 
-  def divideUnsigned(dividend: scala.Long, divisor: scala.Long): scala.Long =
-    divULong(dividend, divisor)
+  @inline def parseUnsignedLong(s: String, radix: Int): scala.Long =
+    undefined
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Long.scala]: Not an intrinsic, needs to be implemented in the library code.

File: javalib/src/main/scala/java/lang/Long.scala [hunk 3]
```diff
@@ -36,81 +48,317 @@ object Long {
...
+        buffer(0) = '-'
+      }
+
+      new String(buffer)
+    }
+  }
 
-  def toHexString(l: scala.Long): String = ???
+  @inline def toUnsignedString(l: scala.Long, radix: Int): String =
+    undefined
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Long.scala]: Not an intrinsic either, all string conversions are library code.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 4]
```diff
@@ -49,40 +62,79 @@ object Integer {
...
+      result = -result
+      if (result < 0) throw new NumberFormatException(s)
+    }
 
-  @inline def toUnsignedLong(x: Int): scala.Long =
-    x.toLong & 0xffffffffL
+    result
+  }
 
   def bitCount(i: scala.Int): scala.Int = {
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: Use [`llvm.ctpop.i32`](http://llvm.org/docs/LangRef.html#llvm-ctpop-intrinsic) here. Also add `@inline`.

File: javalib/src/main/scala/java/lang/Long.scala [hunk 5]
```diff
@@ -36,81 +48,317 @@ object Long {
...
+      if (result < 0) {
+        throw new NumberFormatException(s)
+      }
+    }
+
+    result
+  }
 
-  @inline def toUnsignedString(i: scala.Long, radix: Int): String = ???
+  @inline def bitCount(l: scala.Long): scala.Int = {
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Long.scala]: Use [`llvm.ctpop.i64`](http://llvm.org/docs/LangRef.html#llvm-ctpop-intrinsic) here. Also add `@inline`.

File: javalib/src/main/scala/java/lang/Long.scala (hunk 1 above)
Comment (densh) [inline javalib/src/main/scala/java/lang/Long.scala]: See https://github.com/scala-native/scala-native/pull/210/files?diff=split#diff-ae8efe40d10efb6271e358ad1fb7e0c8R163

File: javalib/src/main/scala/java/lang/Short.scala [hunk 6]
```diff
@@ -41,28 +56,50 @@ object Short {
...
+  @inline def parseShort(s: String, radix: scala.Int): scala.Short = {
+    val i = Integer.parseInt(s, radix)
+    if (i < MIN_VALUE || i > MAX_VALUE)
       throw new NumberFormatException(s"""For input string: "$s"""")
     else
-      r.toShort
+      i.toShort
   }
 
+  @inline def reverseBytes(i: scala.Short): scala.Short =
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Short.scala]: Use `llvm.bswap.i16` here.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 7]
```diff
@@ -49,40 +62,79 @@ object Integer {
...
-
-  @inline def toString(i: scala.Int): String =
-    ???
-
-  @inline def toUnsignedString(i: Int, radix: Int): String =
-    toStringBase(i, radix)
-
-  @inline def compare(x: scala.Int, y: scala.Int): scala.Int =
-    if (x == y) 0 else if (x < y) -1 else 1
+  private def parse(s: String,
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: That's a big method, no need for `@inline`.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 8]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
   }
 
-  @inline def divideUnsigned(dividend: Int, divisor: Int): Int =
+  @inline def compare(x: scala.Int, y: scala.Int): scala.Int =
+    if (x == y) 0 else if (x < y) -1 else 1
+
+  @inline def compareUnsigned(x: scala.Int, y: scala.Int): scala.Int =
+    compare(x ^ scala.Int.MinValue, y ^ scala.Int.MinValue)
+
+  @inline def decode(nm: String): Integer = {
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: That's a big method, no need for `@inline`.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 9]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
+    do {
+      count -= 1
+      buffer(count) = ((k & 7) + '0').toChar
+      k >>>= 3
+    } while (count > 0)
+
+    new String(buffer)
+  }
+
+  @inline def toString(i: scala.Int): String = {
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: No `@inline` here either.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 10]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
+        buffer(last_digit) = (positive_value + '0').toChar
+        last_digit -= 1
+        count = last_digit
+
+        new String(buffer, 0, count)
+      }
+    }
+  }
+
+  @inline def toString(_i: scala.Int, _radix: scala.Int): String = {
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: No `@inline`.

File: javalib/src/main/scala/java/lang/Long.scala [hunk 11]
```diff
@@ -36,81 +48,317 @@ object Long {
   final val SIZE      = 64
   final val BYTES     = 8
 
-  @inline def toString(i: scala.Long, radix: Int): String = ???
+  @inline private def parse(s: String,
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Long.scala]: No `@inline`.

File: javalib/src/main/scala/java/lang/Long.scala [hunk 12]
```diff
@@ -36,81 +48,317 @@ object Long {
...
+    }
 
-  def parseUnsignedLong(s: String, radix: Int): scala.Long = ???
+  @inline def hashCode(value: scala.Long): Int =
+    value.toInt ^ (value >>> 32).toInt
 
-  @inline def parseUnsignedLong(s: String): scala.Long =
-    parseUnsignedLong(s, 10)
+  def highestOneBit(i: scala.Long): scala.Long = {
+    // ToDo reimplement
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Long.scala]: Please don't leave TODO comments in the code.

**Comment by densh** [comment]
Also typo in the commit messages: "Impleme**n**t.."

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 13]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
+        }
+
+        var last_digit = first_digit
+        var quot       = positive_value
+        do {
+          val res = quot / 10
+          var digit_value = quot - ((res << 3) + (res << 1))
+          digit_value += '0'
+          last_digit += 1
+          buffer(last_digit) = digit_value.toChar
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: This part used to be `buffer[last_digit++]` in Harmony, it means `last_digit += 1` should come after `buffer(last_digit) = ...`.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 14]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
+          quot = res
+        } while (quot != 0)
+
+        last_digit -= 1
+        val count = last_digit
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: Here is the same mistake, `last_digit--` means that you first read previous value and then decrement. `val count = last_digit` needs to come before `last_digit -= 1`.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 15]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
+        do {
+          val tmp = buffer(last_digit)
+
+          last_digit -= 1
+          buffer(last_digit) = buffer(first_digit)
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: The same mistake again.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 16]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
+
+          first_digit += 1
+          buffer(first_digit) = tmp
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: The same mistake again.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 17]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
+        var number: Int = 0
+        var start       = false
+        var k           = 0
+        while (k < 9) {
+          count = 0
+          number = decimalScale(k)
+          if (positive_value < number) {
+            if (start) {
+              last_digit += 1
+              buffer(last_digit) = '0'
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: The same mistake as before.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 18]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
+            count += 1
+          }
+
+          if (count > 0 && !start) {
+            start = true
+          }
+
+          if (start) {
+            last_digit += 1
+            buffer(last_digit) = (count + '0').toChar
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: The same mistake.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 19]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
+          }
+
+          k += 1
+        }
+
+        last_digit += 1
+        buffer(last_digit) = (positive_value + '0').toChar
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: The same mistake.

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 20]
```diff
@@ -105,71 +157,420 @@ object Integer {
...
+        last_digit -= 1
+        count = last_digit
```
Comment (densh) [inline javalib/src/main/scala/java/lang/Integer.scala]: The same mistake.

**Comment by densh** [comment]
Please double-check all of the looping code with `x++`/`++x`/`x--`/`--x`. I've explained some issues around that above.

**Comment by densh** [comment]
Needs a rebase on top of latest master due to CI changes.

**Comment by densh** [comment]
Superseded by https://github.com/scala-native/scala-native/issues/238

**Comment by t-hiroyoshi** [comment]
@densh Sorry,  I was sick in bed for a week. I'll do #212

**Comment by densh** [comment]
@t-hiroyoshi No worries, get better! ☕

**Pull request closed without merging**